File: bondly-api/internal/dto/content.go
package dto

import "time"

// CreateContentRequest 创建内容请求结构
type CreateContentRequest struct {
	Title         string  `json:"title" binding:"required" example:"文章标题"`
	Content       string  `json:"content" binding:"required" example:"文章内容"`
	Type          string  `json:"type" binding:"required" example:"article"` // article, post, comment
	Status        string  `json:"status" example:"draft"`                    // draft, published, archived
	CoverImageURL *string `json:"cover_image_url" example:"https://example.com/image.jpg"`
}

// UpdateContentRequest 更新内容请求结构
type UpdateContentRequest struct {
	Title         *string `json:"title" example:"更新后的标题"`
	Content       *string `json:"content" example:"更新后的内容"`
	Type          *string `json:"type" example:"article"`
	Status        *string `json:"status" example:"published"`
	CoverImageURL *string `json:"cover_image_url" example:"https://example.com/new-image.jpg"`
}

// ContentResponse 内容响应结构
type ContentResponse struct {
	ID            int64        `json:"id" example:"1"`
	AuthorID      int64        `json:"author_id" example:"1"`
	Title         string       `json:"title" example:"文章标题"`
	Content       string       `json:"content" example:"文章内容"`
	Type          string       `json:"type" example:"article"`
	Status        string       `json:"status" example:"published"`
	CoverImageURL *string      `json:"cover_image_url" example:"https://example.com/image.jpg"`
	Likes         int64        `json:"likes" example:"10"`
	Dislikes      int64        `json:"dislikes" example:"2"`
	Views         int64        `json:"views" example:"100"`
	CreatedAt     time.Time    `json:"created_at" example:"2023-12-01T10:00:00Z"`
	UpdatedAt     time.Time    `json:"updated_at" example:"2023-12-01T10:00:00Z"`
	Author        UserResponse `json:"author"`
}

// ListContentRequest 获取内容列表请求结构
type ListContentRequest struct {
	Page     int    `form:"page" binding:"min=1" example:"1"`
	Limit    int    `form:"limit" binding:"min=1,max=100" example:"10"`
	Type     string `form:"type" example:"article"`
	Status   string `form:"status" example:"published"`
	AuthorID int64  `form:"author_id" example:"1"`
}

// ListContentResponse 获取内容列表响应结构
type ListContentResponse struct {
	Data       []ContentResponse `json:"data"`
	Pagination PaginationData    `json:"pagination"`
}

// PaginationData 分页数据结构
type PaginationData struct {
	Total      int64 `json:"total" example:"100"`
	Page       int   `json:"page" example:"1"`
	Limit      int   `json:"limit" example:"10"`
	TotalPages int   `json:"total_pages" example:"10"`
}

// UserInteractionStats 用户互动统计
type UserInteractionStats struct {
	UserID         int64 `json:"user_id" example:"1"`
	TotalLikes     int64 `json:"total_likes" example:"100"`
	TotalDislikes  int64 `json:"total_dislikes" example:"10"`
	TotalBookmarks int64 `json:"total_bookmarks" example:"50"`
	TotalShares    int64 `json:"total_shares" example:"25"`
}

// CreateInteractionRequest 创建内容互动请求
type CreateInteractionRequest struct {
	ContentID       int64  `json:"content_id" binding:"required" example:"1"`
	InteractionType string `json:"interaction_type" binding:"required" example:"like"` // like, dislike, bookmark, share
	UserID          int64  `json:"-"`                                                  // 从JWT中获取，不在请求体中
}

// ContentInteraction 内容互动响应
type ContentInteraction struct {
	ID              int64         `json:"id" example:"1"`
	ContentID       int64         `json:"content_id" example:"1"`
	UserID          int64         `json:"user_id" example:"1"`
	InteractionType string        `json:"interaction_type" example:"like"`
	CreatedAt       string        `json:"created_at" example:"2023-12-01T10:00:00Z"`
	User            *UserResponse `json:"user,omitempty"`
}

// InteractionStats 内容互动统计
type InteractionStats struct {
	ContentID        int64                 `json:"content_id" example:"1"`
	Likes            int64                 `json:"likes" example:"10"`
	Dislikes         int64                 `json:"dislikes" example:"2"`
	Bookmarks        int64                 `json:"bookmarks" example:"5"`
	Shares           int64                 `json:"shares" example:"3"`
	UserInteractions UserInteractionStatus `json:"user_interactions"`
}

// UserInteractionStatus 用户互动状态
type UserInteractionStatus struct {
	Liked      bool `json:"liked" example:"true"`
	Disliked   bool `json:"disliked" example:"false"`
	Bookmarked bool `json:"bookmarked" example:"true"`
}

File: bondly-fe/src/api/content.ts
import { apiRequest } from './request';

// 内容类型定义
export interface Content {
  id: number;
  author_id: number;
  title: string;
  content: string;
  type: string;
  status: string;
  cover_image_url?: string;
  nft_token_id?: number;
  nft_contract_address?: string;
  ip_fs_hash?: string;
  metadata_hash?: string;
  likes: number;
  dislikes: number;
  views: number;
  created_at: string;
  updated_at: string;
  author?: {
    id: number;
    nickname: string;
    avatar_url?: string;
    reputation_score?: number;
  };
}

export interface CreateContentRequest {
  title: string;
  content: string;
  type?: string;
  status?: string;
  cover_image_url?: string;
}

export interface UpdateContentRequest {
  title?: string;
  content?: string;
  type?: string;
  status?: string;
  cover_image_url?: string;
  nft_token_id?: number;
  nft_contract_address?: string;
  ip_fs_hash?: string;
  metadata_hash?: string;
}

export interface ContentListResponse {
  contents: Content[];
  pagination: {
    total: number;
    page: number;
    limit: number;
    total_pages: number;
  };
}

// 获取内容列表
export const getContentList = async (params?: {
  page?: number;
  limit?: number;
  status?: string;
  author_id?: number;
  type?: string;
}): Promise<ContentListResponse> => {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.append('page', params.page.toString());
  if (params?.limit) searchParams.append('limit', params.limit.toString());
  if (params?.status) searchParams.append('status', params.status);
  if (params?.author_id) searchParams.append('author_id', params.author_id.toString());
  if (params?.type) searchParams.append('type', params.type);

  const queryString = searchParams.toString();
  const endpoint = `/api/v1/content${queryString ? `?${queryString}` : ''}`;
  
  const response = await apiRequest<ContentListResponse>(endpoint, {
    method: 'GET',
  });
  return response;
};

// 获取单个内容详情
export const getContentById = async (id: number): Promise<Content> => {
  const response = await apiRequest<Content>(`/api/v1/content/${id}`, {
    method: 'GET',
  });
  return response;
};

// 创建内容
export const createContent = async (data: CreateContentRequest): Promise<Content> => {
  const response = await apiRequest<Content>('/api/v1/content', {
    method: 'POST',
    body: JSON.stringify(data),
  });
  return response;
};

// 更新内容
export const updateContent = async (id: number, data: UpdateContentRequest): Promise<Content> => {
  const response = await apiRequest<Content>(`/api/v1/content/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });
  return response;
};

// 删除内容
export const deleteContent = async (id: number): Promise<void> => {
  await apiRequest<void>(`/api/v1/content/${id}`, {
    method: 'DELETE',
  });
};

// 获取用户发布的内容
export const getUserContents = async (userId: number, params?: {
  page?: number;
  limit?: number;
  status?: string;
}): Promise<ContentListResponse> => {
  const searchParams = new URLSearchParams();
  searchParams.append('author_id', userId.toString());
  if (params?.page) searchParams.append('page', params.page.toString());
  if (params?.limit) searchParams.append('limit', params.limit.toString());
  if (params?.status) searchParams.append('status', params.status);

  const endpoint = `/api/v1/content?${searchParams.toString()}`;
  
  const response = await apiRequest<ContentListResponse>(endpoint, {
    method: 'GET',
  });
  return response;
};

// 获取草稿列表
export const getDrafts = async (params?: {
  page?: number;
  limit?: number;
}): Promise<ContentListResponse> => {
  return getContentList({
    ...params,
    status: 'draft'
  });
};

// 获取已发布内容
export const getPublishedContents = async (params?: {
  page?: number;
  limit?: number;
  type?: string;
}): Promise<ContentListResponse> => {
  return getContentList({
    ...params,
    status: 'published'
  });
}; 

// 内容互动相关接口
export interface ContentInteraction {
  id: number;
  content_id: number;
  user_id: number;
  interaction_type: 'like' | 'dislike' | 'bookmark' | 'share';
  created_at: string;
  user?: {
    id: number;
    nickname: string;
    avatar_url?: string;
  };
}

export interface CreateInteractionRequest {
  content_id: number;
  interaction_type: 'like' | 'dislike' | 'bookmark' | 'share';
}

export interface InteractionStats {
  content_id: number;
  likes: number;
  dislikes: number;
  bookmarks: number;
  shares: number;
  user_interactions: {
    liked: boolean;
    disliked: boolean;
    bookmarked: boolean;
  };
}

// 创建内容互动
export const createInteraction = async (data: CreateInteractionRequest): Promise<ContentInteraction> => {
  const response = await apiRequest<ContentInteraction>('/api/v1/content-interactions', {
    method: 'POST',
    body: JSON.stringify(data),
  });
  return response;
};

// 删除内容互动
export const deleteInteraction = async (contentId: number, interactionType: string): Promise<void> => {
  await apiRequest<void>(`/api/v1/content-interactions/${contentId}/${interactionType}`, {
    method: 'DELETE',
  });
};

// 获取内容的互动统计
export const getInteractionStats = async (contentId: number): Promise<InteractionStats> => {
  const response = await apiRequest<InteractionStats>(`/api/v1/content-interactions/${contentId}/stats`, {
    method: 'GET',
  });
  return response;
};

// 获取用户对特定内容的互动状态
export const getUserInteractionStatus = async (contentId: number): Promise<{
  liked: boolean;
  disliked: boolean;
  bookmarked: boolean;
}> => {
  const response = await apiRequest<{
    liked: boolean;
    disliked: boolean;
    bookmarked: boolean;
  }>(`/api/v1/content-interactions/${contentId}/user-status`, {
    method: 'GET',
  });
  return response;
}; 